### PR TITLE
Editor Camera Improved

### PR DESCRIPTION
**Pull request by SuperWangKai**
1. Camera now has a virtual look-at point. Camera can now rotate around the look-at point or as well as itself. No focus will be lost during camera operation (issue #2039). New object can place at this look-at point(view center).

2. Better camera close-look - adjust pos/zoom of camera and make view frustum fit the calculated bounding box of selected nodes and components.

3. Double click component in Hierarchy Window can also look close to the node.

4. Camera smooth interpolation - pos/rot/zoom smooth interpolating for close-look, top/front/side view switching.

5. Removed MouseWheelCameraPosition to make camera less confusing. For ortho camera mode, zoom means the 'zoom' value of the camera will be changed, and for the perspective camera mode, camera distance will be changed. 

6. Key/Mouse changed - Added F key(Standard Hotkey, many popular tools use this key) and NumPad Period(Blender Hotkey) for camera close-look. Added NumPad 9 key for side view (same as NumPad 3, blender has them two). Moving middle mouse button for orbiting around look-at and right button movement for orbiting around camera itself (For standard Hotkey).

7. Some camera related code cleanup.

8. Added 'Rotate Around Selection' for supporting legacy way of camera orbiting

### CODE REVIEW

**Comment by SuperWangKai** [comment]
You can download the bin.zip, backup your original bin folder, extract the downloaded and override for quick test - [bin.zip](https://github.com/SuperWangKai/UrhoTestStorage/blob/master/bin.zip)

File: bin/Data/Scripts/Editor.as [hunk 1]
```diff
@@ -208,6 +208,8 @@ void LoadConfig()
         if (cameraElem.HasAttribute("viewportmode")) viewportMode = cameraElem.GetUInt("viewportmode");
         if (cameraElem.HasAttribute("mouseorbitmode")) mouseOrbitMode = cameraElem.GetInt("mouseorbitmode");
         if (cameraElem.HasAttribute("mmbpan")) mmbPanMode = cameraElem.GetBool("mmbpan");
+        if (cameraElem.HasAttribute("mmbpan")) rotateAroundSelect = cameraElem.GetBool("rotatearoundselect");
```
Comment (1vanK) [inline bin/Data/Scripts/Editor.as]: typo

**Comment by 1vanK** [comment]
1) delete cofig file
2) open editor - middle mouse button  rotate camera arount cener of scene
3) add node - middle mouse button rotate camera arount cener of scene
4) move camera - middle mouse button rotate camera around its own center
5) doubleclick on node - camera moved to unknown position

video http://dropmefiles.com/pWwd1

**Comment by SuperWangKai** [comment]
@1vanK Thank you for the review. Both should be fixed.

**Comment by 1vanK** [comment]
Rotation around some virtual point still does not works after moving a camera. I do not see difference between MMB and RMB

**Comment by SuperWangKai** [comment]
MMB - Rotation around the look-at point, RMB - Rotation around camera center. This works under Standard Hotkey mode. I don't know what do you mean by "does not work after moving a camera". If there is a bug pls show me the steps.

**Comment by 1vanK** [comment]
> If there is a bug pls show me the steps.

I showed it in video

**Comment by SuperWangKai** [comment]
In your video 17s- 19s, you are using RMB to rotate, the camera is orbiting around its own center. Before that, you are using MMB, rotating around the look-at point. The camera behavior is correct, as expected.

**Comment by 1vanK** [comment]
No, a use only MMB

**Comment by 1vanK** [comment]
In this video I use both MMB and RMB, you can see small difference of the point of rotation
http://dropmefiles.com/ccZyY

**Comment by 1vanK** [comment]
Looks like camera use local point instead world coords for origin

**Comment by 1vanK** [comment]
Also pls enable old method as default

**Comment by SuperWangKai** [comment]
Mouse wheel scroll for zoom in out, or you can use WASD.  It's normal and most of the editors use this way. So I will keep it as default.

**Comment by 1vanK** [comment]
WASD move focus points with camera as I show in video.

> Mouse wheel scroll for zoom in out,

Hm. before wheel change fov of camera, how to do it now?

> So I will keep it as default.

Then I vote against this PR

**Comment by SuperWangKai** [comment]
> WASD move focus points with camera as I show in video.
Yes, it's expected. WASD is for First Person movement. I don't want to explain these common practice to you any more, please compare my implementation to any of other popular editors - 3ds max, blender, Unity, UE4 before comment. There is no problem with the orbit or pan or zoom.

> Hm. before wheel change fov of camera, how to do it now?
In the old editor, it's the **zoom** value changed when you scroll your MMB. Now it's well supported in Orthographic mode, by scrolling MMB. I don't see there is any need to have the zoom value changed in perspective mode. If there is in some very special scenario, it should be definitely put in some other place. Plus, the old one is very frustrating - you need to toggle this switch from the editor setting window every single time if you changed your camera from ortho to perspective or vise versa by pressing the Num 5 key, this is silly.

> Then I vote against this PR
Just let others to decide. I will stop my work here if there is no further typo or bug in my implementation  since the right thing have been done. Your opinion is just one vote. PR is for improvements not for keeping the defects (I have my compromise to keep the old broken way). The UX of editor camera is largely improved IMO and if this community don't appreciate my improvements, I'll be happy to keep it for myself.

Too much noise here and I'm bored. Community, please have a look at this.

Thanks.

**Comment by eugeneko** [comment]
> Community, please have a look at this.

**Brief:** I put this PR on hold for now. Neither closed nor merged.

**Long:** I am not ready to review this PR mindfully now, it's just too big. Then, Editor is just getting more and more messy as time passes. I think that Editor should be re-factored, re-architectured and partially moved into C++ classes. So _personally_ I will avoid merging anything big into Editor. When the time comes, I'll return here.

**Comment by weitjong** [comment]
IMHO, we should not stop merging PRs for the editor, especially when they are really good, even if/when someone has plan to rewrite Editor from scratch. I keep hearing people saying the same thing over and over again, but in the end only a few actually made it (or perhaps they made it but decided to keep it for themselves or to capitalize/commercialize it as differentiator with Urho (embrace and extend strategy). So, until I see something more concrete is happening, I will leave the door open.

@SuperWangKai Please be patience and do not close the PR. As you know our current situation, the core devs are still struggling to fill up the role left by cadaver.

**Comment by SuperWangKai** [comment]
I surely will.

**Comment by SuperWangKai** [comment]
I will, surely.

**Comment by SuperWangKai** [comment]
I think since no one will merge review the PR, I would like to close it.

**Comment by weitjong** [comment]
I actually plan to do it at some point in time if no one else done it by then.

**Comment by weitjong** [comment]
I am in the middle of git rebasing at the moment and I am not able to checkout and safe-keep your master branch locally yet. So, please keep those commits there for the next few days. Thanks.

**Comment by weitjong** [comment]
Sorry for the delay. I have a local copy of your master branch now and will review the changes later.

**Comment by SuperWangKai** [comment]
OK, Please let me know if there is anything I need to do.

**Comment by SuperWangKai** [comment]
Hi WeiJong,

I have another small pull request to submit and I wonder if this PR can be closed. If there is any concern pls let me know.

**Comment by weitjong** [comment]
You can close it as I already have your changes cloned locally. Sorry that I haven't found time for reviewing it. However, if what you mean by "closing" it is that you don't want it to be merged anymore then I also respect that.

**Comment by SuperWangKai** [comment]
I am glad that you are working on it. I will keep this PR open and contine my work on another local copy.

**Comment by weitjong** [comment]
Thanks for your patience. I definitely like the improvement over the camera smooth transition effect. Well done. I also have no issue changing the keybinding and mouse button binding proposed in your PR. What concerns me a little, however, is the line get blurred in between the "Blender" vs "Standard" mode. But I don't see how we can clean this out without major rewriting the Editor. Ideally the the editor actions and their binding should be easily customize like most other editors, before talking about having custom theme to make it Blender-like or Urho-standard-like (whatever that is). For now I think we shall just make do of what we have. After you have updated the documentation for the Editor in this page section below, then I think the PR can be merged. 

https://github.com/urho3d/Urho3D/blob/master/Docs/GettingStarted.dox#L809

I found one thing that looks like a bug to me. After double-clicking a component in the Hierarchy window, the camera zoom in on the double-clicked node/component as expected. However, the subsequent WASD keys is being ignored by the Editor. I have to click elsewhere or select something else on the scene first to make WASD works again.

**Comment by SuperWangKai** [comment]
@weitjong 
Thanks for the review and sorry for the delay. I've been busy doing some game engine job from which I pay for the bill :P . I will surely check the issue you found soon and if it's a bug it will be fixed.

I also have also did some experiments on the Editor enhancement: 
https://discourse.urho3d.io/t/quick-expriments-on-editor/3571/2 

and the color picker already worked as good as it should be with some 'reviewable' changes :)  I am glad to contine pushing editor enhancements if they are align with the roadmap in the future.

**Comment by weitjong** [comment]
Re. the bug, I think it could be just a simple expectation from my part. After double-clicking the focus may stay at the Hierarchy Window and so the WASD is ignored.

**Comment by SuperWangKai** [comment]
Yes, the Hierarchy Window focus issue is not introduced by my PR and the original Editor has this too. We could improve it later though.

We can also improve the key bindings and make the interaction better (I have already made some code refactor related to Standard/Blender Hotkey in this PR and it's the main reason why the code review is hard...), however, I doubt if a full Blender key binding implementation is necessary, since Blender also has its own issue. It's an open topic though. Personally, I prefer simple, intuitive, unified way of interaction. 

I have modified the documentation for the Editor, please have a check.

**Comment by weitjong** [comment]
Thanks for that. I will merge this shortly when there is no objections from others.

**Comment by eugeneko** [comment]
I haven't reviewed the code thoughtfully, but I'm absolutely fine with it.

**Pull request merged**